### PR TITLE
only return small set of targets by default from dataset wrapper

### PR DESCRIPTION
**Pull request by pmeier**
From benchmarking, it seems that the detection references with v2 are quite a bit slower than v1 for two reasons:

1. The input sample for COCO is fairly large with roughly > 50 items. The largest part come from the `"segmentations"` key, which stores the vertex coordinates of polygons in nested lists. Recursing through them every time with `pytree` is slow.
2. The `"segmentations"` key will be decoded into a `datapoints.Mask` and later on transformed although for regular object detection, this is not needed.

This PR introduces the `target_keys` parameter that let's users select which keys they actually want.

ToDo:

- [x] add support for `"all"` if users want everything
- [x] document this
- [x] honor `target_keys` in all irregular datasets besides `CocoDetection`

cc @vfdev-5

### CODE REVIEW

File: torchvision/datapoints/_dataset_wrapper.py [hunk 1]
```diff
@@ -288,30 +290,40 @@ def wrapper(idx, sample):
         if not target:
             return image, dict(image_id=image_id)
 
+        spatial_size = tuple(F.get_spatial_size(image))
         batched_target = list_of_dicts_to_dict_of_lists(target)
+        target = {}
 
-        batched_target["image_id"] = image_id
+        if "image_id" in target_keys:
+            target["image_id"] = image_id
```
Comment (NicolasHug) [inline torchvision/datapoints/_dataset_wrapper.py:298]: Instead of adding all those `if` blocks, should we just keep these parts intact and just drop the unnecessary keys at the end?

File: torchvision/datapoints/_dataset_wrapper.py [hunk 2]
```diff
@@ -288,30 +290,40 @@ def wrapper(idx, sample):
...
+
+        if "masks" in target_keys:
+            target["masks"] = datapoints.Mask(
+                torch.stack(
+                    [
+                        segmentation_to_mask(segmentation, spatial_size=spatial_size)
+                        for segmentation in batched_target["segmentation"]
+                    ]
+                ),
+            )
```
Comment (NicolasHug) [inline torchvision/datapoints/_dataset_wrapper.py:318]: If we don't do that by default anymore, is this going to bias the benchmarks in favour of v2 (because the v1 path still does it no matter what we pass)?

Comment (pmeier) [reply torchvision/datapoints/_dataset_wrapper.py:318]: We can setup the benchmarks to always decode the masks and drop them after. With this we have comparable results, while users still get the best performance.

File: torchvision/datapoints/_dataset_wrapper.py (hunk 1 above)
Comment (pmeier) [reply torchvision/datapoints/_dataset_wrapper.py:298]: Somewhat a trade-off. If we leave everything intact until the very end, we pay for the in between, e.g. decode masks just to throw them away shortly after. I don't have a good solution for this yet. It would also need to factor in the `target_types="all"` case.

File: test/datasets_utils.py [hunk 3]
```diff
@@ -572,9 +572,22 @@ def test_transforms_v2_wrapper(self, config):
...
+                    for item in flat_sample:
+                        if isinstance(item, (Datapoint, PIL.Image.Image)):
+                            break
+                        elif isinstance(item, torch.Tensor):
+                            raise AssertionError(
+                                "Found a plain tensor before an explicit image or video. "
+                                "This is not compatible with the tensor fallback heuristic."
+                            )
+                    else:
+                        raise AssertionError("Found no image or video in sample.")
```
Comment (pmeier) [inline test/datasets_utils.py:590]: Driveby: instead of just checking that we have an explicit image, we now also check that no plain tensor comes before it.

File: test/test_datasets.py [hunk 4]
```diff
@@ -771,6 +771,8 @@ def _create_annotations(self, image_ids, num_annotations_per_image):
                     bbox=torch.rand(4).tolist(),
                     segmentation=[torch.rand(8).tolist()],
                     category_id=int(torch.randint(91, ())),
+                    area=float(torch.rand(1)),
+                    iscrowd=int(torch.randint(2, size=(1,))),
```
Comment (pmeier) [inline test/test_datasets.py:775]: We need the full sample now for `target_keys="all"`.

File: torchvision/datapoints/_dataset_wrapper.py [hunk 5]
```diff
@@ -78,8 +82,21 @@ def wrap_dataset_for_transforms_v2(dataset):
...
-    return VisionDatasetDatapointWrapper(dataset)
+    if not (
+        target_keys is None
+        or target_keys == "all"
+        or (isinstance(target_keys, collections.abc.Collection) and all(isinstance(key, str) for key in target_keys))
+    ):
+        raise ValueError(
+            f"`target_keys` can be None, 'all', or a collection of strings denoting the keys to be returned, "
+            f"but got {target_keys}"
+        )
```
Comment (NicolasHug) [inline torchvision/datapoints/_dataset_wrapper.py:97]: Do we also want to raise an error if `target_keys` is passed for any dataset that isn't `Coco`, `WIDERFace`, etc. ?

File: test/datasets_utils.py (hunk 3 above)
Comment (pmeier) [reply test/datasets_utils.py:590]: Ah shoot my bad. You are right. Let me revert.

File: torchvision/datapoints/_dataset_wrapper.py [hunk 6]
```diff
@@ -123,7 +140,8 @@ def __init__(self, dataset):
                 raise TypeError(msg)
 
         self._dataset = dataset
-        self._wrapper = wrapper_factory(dataset)
+        self._target_keys = target_keys
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/datapoints/_dataset_wrapper.py:143]: Nope, can remove.

File: torchvision/datapoints/_dataset_wrapper.py [hunk 7]
```diff
@@ -181,6 +199,26 @@ def pil_image_to_mask(pil_image):
     return datapoints.Mask(pil_image)
 
 
+def parse_target_keys(*, available, default):
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/datapoints/_dataset_wrapper.py:202]: Can do.

File: torchvision/datapoints/_dataset_wrapper.py (hunk 5 above)
Comment (pmeier) [reply torchvision/datapoints/_dataset_wrapper.py:97]: Yeah, let's play safe here.

**Comment by pmeier** [comment]
Sneak peak into the performance gains of this change

```
Summaries

           v2 / v1
Tensor        0.32
PIL           0.33

                     [a]   [b]   [c]   [d]   [e]
   Tensor, v1, [a]  1.00  3.16  0.97  2.91  2.79
   Tensor, v2, [b]  0.32  1.00  0.31  0.92  0.88
      PIL, v1, [c]  1.03  3.26  1.00  2.99  2.87
      PIL, v2, [d]  0.34  1.09  0.33  1.00  0.96
Datapoint, v2, [e]  0.36  1.14  0.35  1.04  1.00

Slowdown as row / col
```

<details>
  <summary>Details</summary>

  ```
  ############################################################
  detection-ssdlite
  ############################################################
  input_type='Tensor', api_version='v1'
  
  Results computed for 1_000 samples
  
                            median          std   
  ConvertCocoPolysToMaskV1    2663 µs +-   4269 µs
  PILToTensorV1                251 µs +-     71 µs
  RandomIoUCropV1              456 µs +-   6950 µs
  RandomHorizontalFlipV1        17 µs +-    200 µs
  ConvertImageDtypeV1          271 µs +-    172 µs
  
  total                       3657 µs
  ------------------------------------------------------------
  input_type='Tensor', api_version='v2'
  
  Results computed for 1_000 samples
  
                                 median          std   
  WrapCocoSampleForTransformsV2      61 µs +-     34 µs
  PILToTensor                       218 µs +-     46 µs
  RandomIoUCrop                     394 µs +-   6302 µs
  RandomHorizontalFlip              132 µs +-    116 µs
  ConvertDtype                      178 µs +-    128 µs
  SanitizeBoundingBox               173 µs +-     14 µs
  
  total                            1156 µs
  ------------------------------------------------------------
  input_type='PIL', api_version='v1'
  
  Results computed for 1_000 samples
  
                            median          std   
  ConvertCocoPolysToMaskV1    2735 µs +-   4329 µs
  RandomIoUCropV1              548 µs +-   7064 µs
  RandomHorizontalFlipV1        19 µs +-    223 µs
  PILToTensorV1                182 µs +-    118 µs
  ConvertImageDtypeV1          282 µs +-    178 µs
  
  total                       3767 µs
  ------------------------------------------------------------
  input_type='PIL', api_version='v2'
  
  Results computed for 1_000 samples
  
                                 median          std   
  WrapCocoSampleForTransformsV2      64 µs +-     26 µs
  RandomIoUCrop                     489 µs +-   6620 µs
  RandomHorizontalFlip              143 µs +-    126 µs
  PILToTensor                       189 µs +-     84 µs
  ConvertDtype                      191 µs +-    121 µs
  SanitizeBoundingBox               182 µs +-     16 µs
  
  total                            1259 µs
  ------------------------------------------------------------
  input_type='Datapoint', api_version='v2'
  
  Results computed for 1_000 samples
  
                                 median          std   
  WrapCocoSampleForTransformsV2      65 µs +-     21 µs
  ToImageTensor                     272 µs +-     92 µs
  RandomIoUCrop                     438 µs +-   6962 µs
  RandomHorizontalFlip              148 µs +-    183 µs
  ConvertDtype                      205 µs +-    162 µs
  SanitizeBoundingBox               185 µs +-     25 µs
  
  total                            1313 µs
  ```

</details>

Roughly 3x over v1. To be fair, v1 is affected by #7489 so this comparison is not totally fair. However, it accurately reflects the current state of our references vs. what can be achieved with v2.
